### ISSUE TITLE
Metrics for MOSEI Emotion Labels

### ISSUE THREAD

**Issue by codeislife99** [A2Zadeh/CMU-MultimodalDataSDK#31]
I was wondering what are the different metrics that are considered for evaluation. One obvious metric which comes to mind is the MSE. What are the others given the distribution of emotional intensities for this dataset is mostly(~92%+) between 0 and 1?

**Comment by A2Zadeh** [comment]
There are various different metrics that can be defined. I personally like regressions since it uses no proxy metric. For example accuracy is a proxy metric which we can _mathematically_  define boundaries but those boundaries have no grounding in behavioral studies. In CMU-MOSEI we did the following: For "x" emotions binary we considered if 0 then nothing else x (for example x could be happy). For multiclass we round the number to the nearest integer and call that a class [0,1,2,3]. Again, my favorite metric is MSE or MAE, but since those are not very intuitive we defined the above proxy.

**Comment by roryjbeard** [comment]
Hi Amir, I found parts of you answer above to be difficult to understand. When you say "In CMU-MOSEI we did the following", are you talking about how you personally scored your own modelling attempts, or how you score competition submissions? For the multi-class rounding example above, are you saying that this is what you recommend should be done if we wish to train using a multi-label cross-entropy loss? Or is it what we should do to have an interpretable score, having *trained* using a regression loss? If it's the latter, then how do you define the score given ground truth labels in {0,1,2,3} and predicted labels in {0,1,2,3}? E.g. one could use micro or macro or weighted F1 score. 'Weighted" weights by support (so Happiness would dominate in contribution), but it doesn't weight by "distance from truth" e.g. gt=3, pred=0 has same effect as gt=3, pred=2, even though the second prediction was "closer".
Moreover, computing a F1 score on a multi-class, multi-label problem could proceed with either (i) binary thresholding each output dimension (e.g. Happiness >0.5 --> Happines=True), or (ii) round the predictions to be in {0,1,2,3,4} and consider any predicted number != ground truth to be a False Negative.

Knowing how you have chosen to deal with these scoring possibilities has a huge effect on the data science (as opposed to modelling) effort of people who use this dataset, so any further clarifications would be greatly appreciated. Thanks!

**Comment by A2Zadeh** [comment]
Thanks for the follow up @roryjbeard. To clarify the original answer, I personally like regression metrics. They align better with intuition. Yes the metrics such as multiclass classification penalizes the labels the same way but underneath they also use soft losses (since the classification loss is not differentiable directly) and those losses most of the time break down to some form of regression. Now let's assume for a second that regression is the best way to go for these scores, intuitively regression are not easy to understand. Therefore we use some proxy metric such as multiclass classification or binary classification. In our work we always try to use the regression metrics to train the models and during testing we also report the proxy metrics too. Did I answer your question?

**Comment by roryjbeard** [comment]
Thanks Amir, that partially clarifies things, but my main question was about how to apply the post-training scoring to the regression values. Knowing which proxy is chosen for final scoring and any subtle details (e.g. how class-specific scores are weighted based on support, whether binary thresholding is performed, at what threshold value, or is it some non-binary scoring method that doesn't throw away the 4-way classification {0,1,2,3} of the original ground truth)? The choice will have drastic effects on the perceived performance, especially given the strong class imbalance of the dataset.
Can you clarify the process for converting real values from the regression to the scores reported in e.g. this paper: https://www.ml.cmu.edu/research/dap-papers/S18/dap-liang-paul-pu.pdf

**Comment by KuzmaNg** [comment]
Hi. I have the same question as @roryjbeard . Is the emotion experiment from Paper(dap-liang-paul-pu) conducted under binary classification or a 4-way classification? I am confused because some types of emotion may not have value-3 samples. And how is the weighted accuracy calculated? Also the F-score, is it only calculated for positive example or the average of both positive and negative? Thanks.
